From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] include: add the DirectShow automation collection interfaces to control.idl

Add IAMCollection, IMediaTypeInfo, IFilterInfo, IRegFilterInfo and IPinInfo
to the QuartzTypeLib in control.idl (purely additive, IIDs 56a868b9..bd as
documented for the DirectShow control interfaces). The dependent headers
(control.h, the strmiids IIDs and quartz's type library) regenerate from it.
---
diff --git a/include/control.idl b/include/control.idl
index 265444a..7cddee2 100644
--- a/include/control.idl
+++ b/include/control.idl
@@ -250,4 +250,97 @@ library QuartzTypeLib
         HRESULT CanSeekBackward( [out, retval] LONG *pCanSeekBackward );
     }
 
+    /***************************************************************************
+     * IAMCollection interface
+     */
+    [
+        object,
+        uuid(56a868b9-0ad4-11ce-b03a-0020af0ba770),
+        oleautomation,
+        dual,
+    ]
+    interface IAMCollection : IDispatch
+    {
+        [propget, id(1)] HRESULT Count( [out, retval] LONG *plCount );
+        [id(0)] HRESULT Item( [in] LONG lItem, [out] IUnknown **ppUnk );
+        [propget, restricted, id(-4)] HRESULT _NewEnum( [out, retval] IUnknown **ppUnk );
+    }
+
+    /***************************************************************************
+     * IMediaTypeInfo interface
+     */
+    [
+        object,
+        uuid(56a868bc-0ad4-11ce-b03a-0020af0ba770),
+        oleautomation,
+        dual,
+    ]
+    interface IMediaTypeInfo : IDispatch
+    {
+        [propget] HRESULT Type( [out, retval] BSTR *strType );
+        [propget] HRESULT Subtype( [out, retval] BSTR *strType );
+    }
+
+    /***************************************************************************
+     * IFilterInfo interface
+     */
+    [
+        object,
+        uuid(56a868ba-0ad4-11ce-b03a-0020af0ba770),
+        oleautomation,
+        dual,
+    ]
+    interface IFilterInfo : IDispatch
+    {
+        HRESULT FindPin( [in] BSTR strPinID, [out] IDispatch **ppUnk );
+        [propget] HRESULT Name( [out, retval] BSTR *strName );
+        [propget] HRESULT VendorInfo( [out, retval] BSTR *strVendorInfo );
+        [propget] HRESULT Filter( [out, retval] IUnknown **ppUnk );
+        [propget] HRESULT Pins( [out, retval] IDispatch **ppUnk );
+        [propget] HRESULT IsFileSource( [out, retval] LONG *pbIsSource );
+        [propget] HRESULT Filename( [out, retval] BSTR *pstrFilename );
+        [propput] HRESULT Filename( [in] BSTR strFilename );
+    }
+
+    /***************************************************************************
+     * IRegFilterInfo interface
+     */
+    [
+        object,
+        uuid(56a868bb-0ad4-11ce-b03a-0020af0ba770),
+        oleautomation,
+        dual,
+    ]
+    interface IRegFilterInfo : IDispatch
+    {
+        [propget] HRESULT Name( [out, retval] BSTR *strName );
+        HRESULT Filter( [out] IDispatch **ppUnk );
+    }
+
+    /***************************************************************************
+     * IPinInfo interface
+     */
+    [
+        object,
+        uuid(56a868bd-0ad4-11ce-b03a-0020af0ba770),
+        oleautomation,
+        dual,
+    ]
+    interface IPinInfo : IDispatch
+    {
+        [propget] HRESULT Pin( [out, retval] IUnknown **ppUnk );
+        [propget] HRESULT ConnectedTo( [out, retval] IDispatch **ppUnk );
+        [propget] HRESULT ConnectionMediaType( [out, retval] IDispatch **ppUnk );
+        [propget] HRESULT FilterInfo( [out, retval] IDispatch **ppUnk );
+        [propget] HRESULT Name( [out, retval] BSTR *ppUnk );
+        [propget] HRESULT Direction( [out, retval] LONG *ppDirection );
+        [propget] HRESULT PinID( [out, retval] BSTR *strPinID );
+        [propget] HRESULT MediaTypes( [out, retval] IDispatch **ppUnk );
+        HRESULT Connect( [in] IUnknown *pPin );
+        HRESULT ConnectDirect( [in] IUnknown *pPin );
+        HRESULT ConnectWithType( [in] IUnknown *pPin, [in] IDispatch *pMediaType );
+        HRESULT Disconnect();
+        HRESULT Render();
+    }
+
 } /* library QuartzTypeLib */
